internal/controller/machineaccount_controller.go

```diff
@@ -0,0 +1,256 @@
+	log := logf.FromContext(ctx).WithName("machineaccount-finalizer")
+	log.Info("Starting finalization")
+
+	// Type assertion to get the machine account object.
+	machineAccount, ok := obj.(*iammiloapiscomv1alpha1.MachineAccount)
+	if !ok {
+		err := fmt.Errorf("unexpected object type %T, expected MachineAccount", obj)
+		log.Error(err, "Type assertion failed")
+		return finalizer.Result{}, err
+	}
+
+	log.Info("Checking if machine account exists in Zitadel", "username", machineAccount.GetUID())
+	_, err := f.Zitadel.GetUser(ctx, string(machineAccount.GetUID()))
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Error(err, "Failed to get user from Zitadel", "userName", machineAccount.GetUID())
+			return finalizer.Result{}, err
+		}
+		log.Info("Machine account not found in Zitadel, nothing to clean up", "uid", machineAccount.GetUID())
+		return finalizer.Result{}, nil
+	}
+
+	log.Info("Deleting machine account from Zitadel", "userName", machineAccount.GetUID())
+	err = f.Zitadel.DeleteUser(ctx, string(machineAccount.GetUID()))
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.Error(err, "Failed to delete user from Zitadel", "username", machineAccount.GetUID())
+			return finalizer.Result{}, err
+		}
+		log.Info("Machine account already deleted from Zitadel", "username", machineAccount.GetUID())
+	}
+
+	log.Info("Successfully finalized machine account", "username", machineAccount.GetUID())
+	return finalizer.Result{}, nil
+}
+
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=machineaccounts,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=machineaccounts/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=iam.miloapis.com,resources=machineaccounts/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the MachineAccount object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.21.0/pkg/reconcile
+func (r *MachineAccountController) Reconcile(ctx context.Context, req mcreconcile.Request) (ctrl.Result, error) {
+	log := logf.FromContext(ctx).WithName("machineaccount-reconciler")
+	log.Info("Starting reconciliation", "request", req)
+
+	machineAccount := &iammiloapiscomv1alpha1.MachineAccount{}
+	err := r.Client.Get(ctx, req.NamespacedName, machineAccount)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("MachineAccount resource not found")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "Failed to get MachineAccount resource")
+		return ctrl.Result{}, err
+	}
```

Comment:
Super minor thing, but you may find it easier to read this by using an if / else if than adding another layer of indentation. 

```suggestion
	err := r.Client.Get(ctx, req.NamespacedName, machineAccount)
	if errors.IsNotFound(err) {
			log.Info("MachineAccount resource not found")
			return ctrl.Result{}, nil
	} else if err != nil {	
		log.Error(err, "Failed to get MachineAccount resource")
		return ctrl.Result{}, err
	}
```

Reply:
that's a great feedback. Updated the whole controller using the same pattern.